BUG Ensure dimensions are cached

tractorcow opened the pull request:
Fixes #47 

I'll do some benchmarking now, but it seems to work if I run a bunch of pages, and then throw an Exception before loading assets from stream (to prove the image isn't loaded by the Image_Backend).

@sminnee do you want to review?

@dhensby I've refactored loadFromContainer so it doesn't immediately load the file into memory. Instead we wait until absolutely necessary.

sminnee commented on src/InterventionBackend.php:
```diff
@@ -144,10 +154,32 @@ public function setImageManager($manager)
      */
     public function loadFromContainer(AssetContainer $assetContainer)
```
Should this method be deprecated now? It seems like clone of setAssetContainer.

tractorcow replied:
It's a part of Image_Backend interface I can't remove it.

sminnee replied:
Didn't we invent Image_Backend, though?

tractorcow replied:
I think removing it from the interface counts as API removal.

sminnee commented on src/InterventionBackend.php:
```diff
@@ -294,27 +324,102 @@ public function getQuality()
+        // cache those anyway.
+        $resource = $this->image;
+        if ($resource) {
+            return [
```
I'd use `getSize()` here, which returns a `Intervention\Image\Size` object with `->width` and `->height` public params. It looks like the underlying code will be called twice otherwise.

Note that `getSize()` is a magic method implemented by one of the GetSizeCommand classes in intervention.

tractorcow replied:
Size is a bit hard to serialize, since it has a pivot as well. I can look at it though.

sminnee replied:
I just meant call once and convert to an array:

```php
$size = $resource->getSize();
return [ $size->width, $size->height ];
```

tractorcow replied:
Oh I see, ok. :D

tractorcow replied:
ok, done and pushed up.

sminnee commented:
Note that this fixes https://github.com/silverstripe/silverstripe-asset-admin/issues/536

tractorcow commented:
Updated with suggested fix.

sminnee commented on src/InterventionBackend.php:
```diff
@@ -294,27 +325,111 @@ public function getQuality()
         $resource = $this->getImageResource();
         if ($resource) {
-            return $resource->getHeight();
+            $result = [
```
use getResourceDimensions() here too

sminnee commented on src/InterventionBackend.php:
```diff
@@ -294,27 +325,111 @@ public function getQuality()
+            return $result;
+        }
+
+        // Check cache for unloded image
```
The cache should probably be checked first

sminnee replied:
Specifically, because the image can't change without the SHA also changing, the cases in which you'd want to discard the cache are rare enough that they can be handled with an explicit cache purge.

This means that the other logic isn't used unnecessarily: the normal flow will be that these fields are warmed on upload and then fetched from the cache for the rest of the object's lifetime.

tractorcow replied:
You can't generate the key if the image has been resized, since the original tuple has been modified (and you can't get the updated "Variant" from anywhere).

tractorcow replied:
> Specifically, because the image can't change without the SHA also changing

The SHA of the tuple is always the SHA of the original, not the SHA of the resized variant.

tractorcow replied:
To put your fears at ease, even though we don't load from cache here, we CAN assume that these values are written to the cache in writeToStore().

The next time this thumbnail is loaded into InterventionBackend (another page load after the initial generation) the cache will be warmed,  $this->image is null, and cache should kick in.

sminnee replied:
Ah yep I see, all good :+1:

tractorcow replied:
This is what I attempted to convey in the code comment.

```php
// We can only cache dimensions from the source file (before any manipulations are performed),
// so ensure that the original image is un-loaded into memory.
// If a manipulation has been performed, we will return the dimensions from the already loaded
// image resource. In any case the 'variant' for these manipulations is unknown, so we couldn't
// cache those anyway.
```

tractorcow commented:
I've added the extra getResourceDimensions, fixed a typo, and improved the documentation to help explain why we seemingly incorrectly forget to query the cache. :)

tractorcow commented:
Docs are now:
```
// If we have a resource already loaded, this means we have modified the resource since the
// original image was loaded. This means the "Variant" tuple key is out of date, and we don't
// have a reliable cache key to load from, or save to. If we use the original tuple as a key,
// we would run the risk of overwriting the original dimensions in the cache, with the values
// of the resized instead.
// Instead, we use the immediately available dimensions attached to this resource, and we will
// rely on cache warming in writeToStore to save these values, where the "Variant" becomes available,
// before the next time this variant is loaded into memory.
```

tractorcow commented:
@dhensby requesting a review please :)
